api/turing/service/ensembling_job_service.go

```diff
@@ -0,0 +1,148 @@
+package service
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/gojek/turing/api/turing/models"
+	"github.com/jinzhu/gorm"
+)
+
+const (
+	sparkHomeFolder string = "/home/spark"
+)
+
+// EnsemblingJobFindByIDOptions contains the options allowed when finding ensembling jobs.
+type EnsemblingJobFindByIDOptions struct {
+	ProjectID *models.ID
+}
+
+// EnsemblingJobListOptions holds query parameters for EnsemblersService.List method.
+type EnsemblingJobListOptions struct {
+	PaginationOptions
+	ProjectID *models.ID
+	Status    models.Status
+}
+
+// EnsemblingJobService is the data access object for the EnsemblingJob from the db.
+type EnsemblingJobService interface {
+	Save(ensembler *models.EnsemblingJob) error
+	FindByID(id models.ID, options EnsemblingJobFindByIDOptions) (*models.EnsemblingJob, error)
+	List(options EnsemblingJobListOptions) (*PaginatedResults, error)
+	GetDefaultEnvironment() string
+	GenerateDefaultJobName(ensemblerName string) string
+	GetEnsemblerDirectory(ensembler models.EnsemblerLike) (string, error)
+	GetArtifactURI(ensembler models.EnsemblerLike) (string, error)
```

Comment:
I'm not onboard with the structure of this service. 
1. `GenerateDefaultJobName` `GetEnsemblerDirectory ` and `GetArtifactURI ` are probably only required in the context of ensembler creation. Hence, you don't need them to be defined at the service level, but instead, you would probably need `CreatePredictionJob` method that incapsulates all this logic.
2. Ensembling job can only work with pyfunc ensemblers, so it doesn't make sense to define the signature of these methods as `(ensembler models.EnsemblerLike)`, but instead it should accept `ensembler models.PyfuncEnsembler` and type check (and casting) should be done at the controller level (and relevant `BadRequest` response should be sent if user is trying to submit a job with non-pyfunc ensembler)

Reply:
Agreed, It's starting to come to me now how this repo is structured. Have implemented the suggestions.